src/Tools/dotnet-monitor/DiagnosticsMonitorCommandHandler.cs

```diff
@@ -144,133 +150,156 @@ public static IHostBuilder CreateHostBuilder(IConsole console, string[] urls, st
+                    services.ConfigureEgress(context.Configuration);
+                    services.ConfigureMetrics(context.Configuration);
+                    services.ConfigureStorage(context.Configuration);
+                    services.ConfigureDefaultProcess(context.Configuration);
                 });
+            }
+
+            //Note this is necessary for config only because Kestrel configuration
+            //is not added until WebHostDefaults are added.
+            hostBuilder.ConfigureWebHostDefaults(webBuilder =>
+            {
+                AddressListenResults listenResults = new AddressListenResults();
+                webBuilder.ConfigureServices(services =>
+                {
+                    services.AddSingleton(listenResults);
+                })
+                .ConfigureKestrel((context, options) =>
+                {
+                    //Note our priorities for hosting urls don't match the default behavior.
+                    //Default Kestrel behavior priority
+                    //1) ConfigureKestrel settings
+                    //2) Command line arguments (--urls)
+                    //3) Environment variables (ASPNETCORE_URLS, then DOTNETCORE_URLS)
+
+                    //Our precedence
+                    //1) Environment variables (ASPNETCORE_URLS, DotnetMonitor_Metrics__Endpoints)
+                    //2) Command line arguments (these have defaults) --urls, --metricUrls
+                    //3) ConfigureKestrel is used for fine control of the server, but honors the first two configurations.
+
+                    string hostingUrl = context.Configuration.GetValue<string>(WebHostDefaults.ServerUrlsKey);
+                    urls = ConfigurationHelper.SplitValue(hostingUrl);
+
+                    var metricsOptions = new MetricsOptions();
+                    context.Configuration.Bind(ConfigurationKeys.Metrics, metricsOptions);
+
+                    //Workaround for lack of default certificate. See https://github.com/dotnet/aspnetcore/issues/28120
+                    options.Configure(context.Configuration.GetSection("Kestrel")).Load();
+
+                    //By default, we bind to https for sensitive data (such as dumps and traces) and bind http for
+                    //non-sensitive data such as metrics. We may be missing a certificate for https binding. We want to continue with the
+                    //http binding in that scenario.
+                    listenResults.Listen(
+                        options,
+                        urls,
+                        metricsOptions.Enabled.GetValueOrDefault(MetricsOptionsDefaults.Enabled) ? metricUrls : Array.Empty<string>());
+                })
+                .UseStartup<Startup>();
+            });
+
             return hostBuilder;
         }
 
+        private static void ConfigureTempApiHashKey(IConfigurationBuilder builder, AuthOptions authenticationOptions)
+        {
+            const string DefaultHashAlgorithm = "SHA512";
```

Comment:
The generatekey command on dotnet-monitor generates a SHA256 key. I think these should be consistent. Additionally, I think SHA256 should be the default; if I recall correctly, there might be a bug in the ApiKeyAuthenticationHandler that assumes the data is only 32 bytes long.

Reply:
refactored to a single generator